File: src/main/java/org/strassburger/lifestealz/commands/SettingsCommand.java
package org.strassburger.lifestealz.commands;

import net.kyori.adventure.text.Component;
import org.bukkit.Material;
import org.bukkit.OfflinePlayer;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.strassburger.lifestealz.LifeStealZ;
import org.strassburger.lifestealz.util.MessageUtils;
import org.strassburger.lifestealz.util.customitems.CustomItem;
import org.strassburger.lifestealz.util.customitems.CustomItemManager;
import org.strassburger.lifestealz.util.geysermc.GeyserManager;
import org.strassburger.lifestealz.util.storage.PlayerData;
import org.strassburger.lifestealz.util.storage.Storage;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.UUID;

public class SettingsCommand implements CommandExecutor, TabCompleter {
    private final LifeStealZ plugin;
    private final FileConfiguration config;
    private final Storage storage;
    private final GeyserManager geyserManager;


    public SettingsCommand(LifeStealZ plugin) {
        this.plugin = plugin;
        this.config = plugin.getConfig();
        this.storage = plugin.getStorage();
        this.geyserManager = plugin.getGeyserManager();
    }

    @Override
    public boolean onCommand(@NotNull CommandSender sender, @NotNull Command command, @NotNull String label, String[] args) {

        if (args.length == 0) {
            sendVersionMessage(sender);
            return false;
        }

        String optionOne = args[0];

        switch (optionOne) {
            case "reload":
                return handleReload(sender);
            case "help":
                return handleHelp(sender);
            case "recipe":
                return handleRecipe(sender, args);
            case "hearts":
                return handleHearts(sender, args);
            case "giveItem":
                return handleGiveItem(sender, args);
            case "data":
                return handleData(sender, args);
            case "dev":
                return handleDev(sender, args);
            default:
                return false;
        }
    }

    private void sendVersionMessage(CommandSender sender) {
        sender.sendMessage(MessageUtils.getAndFormatMsg(true, "messages.versionMsg", "FALLBACK&7You are using version %version%",
                new MessageUtils.Replaceable("%version%", plugin.getDescription().getVersion())));
    }

    private boolean handleReload(CommandSender sender) {
        if (!sender.hasPermission("lifestealz.admin.reload")) {
            throwPermissionError(sender);
            return false;
        }

        plugin.reloadConfig();
        plugin.getLanguageManager().reload();
        plugin.getRecipeManager().registerRecipes();
        sender.sendMessage(MessageUtils.getAndFormatMsg(true, "messages.reloadMsg", "&7Successfully reloaded the plugin!"));
        return true;
    }

    private boolean handleHelp(CommandSender sender) {
        if (!sender.hasPermission("lifestealz.help")) {
            throwPermissionError(sender);
            return false;
        }

        StringBuilder helpMessage = new StringBuilder("<reset><!i><!b> \n&8----------------------------------------------------\n&c&lLifeStealZ &7help page<!b>\n&8----------------------------------------------------\n");
        addHelpEntry(helpMessage, sender, "lifestealz.admin.reload", "/lifestealz reload", "- reload the config");
        addHelpEntry(helpMessage, sender, "lifestealz.admin.setlife", "/lifestealz hearts", "- modify how many hearts a player has");
        addHelpEntry(helpMessage, sender, "lifestealz.admin.giveitem", "/lifestealz giveItem", "- give other players custom items, such as hearts");
        addHelpEntry(helpMessage, sender, "lifestealz.viewrecipes", "/lifestealz recipe", "- view all recipes");
        addHelpEntry(helpMessage, sender, "lifestealz.admin.revive", "/revive", "- revive a player without a revive item");
        addHelpEntry(helpMessage, sender, "lifestealz.admin.eliminate", "/eliminate", "- eliminate a player");
        addHelpEntry(helpMessage, sender, "lifestealz.withdraw", "/withdrawheart", "- withdraw a heart");
        helpMessage.append("\n&8----------------------------------------------------\n<reset><!i><!b> ");

        sender.sendMessage(MessageUtils.formatMsg(helpMessage.toString()));
        return true;
    }

    private void addHelpEntry(StringBuilder helpMessage, CommandSender sender, String permission, String command, String description) {
        if (sender.hasPermission(permission)) {
            helpMessage.append("&c<click:SUGGEST_COMMAND:").append(command).append(">").append(command).append("</click> &8- &7").append(description).append("\n");
        }
    }

    private boolean handleRecipe(CommandSender sender, String[] args) {
        if (!sender.hasPermission("lifestealz.viewrecipes")) {
            throwPermissionError(sender);
            return false;
        }

        if (!(sender instanceof Player)) return false;

        if (args.length < 2) {
            throwUsageError(sender, "/lifestealz recipe <" + String.join(" | ", plugin.getRecipeManager().getRecipeIds()) + ">");
            return false;
        }

        String recipe = args[1];

        if (recipe == null || !plugin.getRecipeManager().getRecipeIds().contains(recipe)) {
            throwUsageError(sender, "/lifestealz recipe <" + String.join(" | ", plugin.getRecipeManager().getRecipeIds()) + ">");
            return false;
        }

        if (!plugin.getRecipeManager().isCraftable(recipe)) {
            sender.sendMessage(MessageUtils.getAndFormatMsg(false, "messages.recipeNotCraftable", "&cThis item is not craftable!"));
            return false;
        }

        plugin.getRecipeManager().renderRecipe((Player) sender, recipe);
        return true;
    }

    private boolean handleHearts(CommandSender sender, String[] args) {
        if (!sender.hasPermission("lifestealz.admin.setlife")) {
            throwPermissionError(sender);
            return false;
        }

        if (args.length < 3) {
            throwUsageError(sender, "/lifestealz hearts <add | set | remove> <player> [amount]");
            return false;
        }

        String optionTwo = args[1];
        List<String> possibleOptionTwo = List.of("add", "set", "remove", "get");

        if (optionTwo == null || !possibleOptionTwo.contains(optionTwo)) {
            throwUsageError(sender, "/lifestealz hearts <add | set | remove> <player> [amount]");
            return false;
        }

        String targetPlayerName = args[2];

        if (targetPlayerName == null) {
            throwUsageError(sender, "/lifestealz hearts <add | set | remove> <player> [amount]");
            return false;
        }

        PlayerData targetPlayerData = null;

        if(plugin.hasGeyser() && plugin.getGeyserPlayerFile().isPlayerStored(targetPlayerName)) {
            targetPlayerData = storage.load(geyserManager.getOfflineBedrockPlayerUniqueId(targetPlayerName));
        } else {
            Player onlinePlayer = plugin.getServer().getPlayer(targetPlayerName);
            UUID targetUUID = (onlinePlayer != null) ? onlinePlayer.getUniqueId() : plugin.getServer().getOfflinePlayer(targetPlayerName).getUniqueId();
            OfflinePlayer targetPlayer = plugin.getServer().getOfflinePlayer(targetUUID);

            if (targetPlayer.getName() == null) {
                throwUsageError(sender, "/lifestealz hearts <add | set | remove> <player> [amount]");
                return false;
            }

            targetPlayerData = storage.load(targetPlayer.getUniqueId());
        }

        Player targetPlayer = plugin.getServer().getPlayer(targetPlayerName);

        if (targetPlayerData == null) {
            sender.sendMessage(MessageUtils.getAndFormatMsg(false, "messages.playerNotFound", "&cPlayer not found!"));
            return false;
        }

        if (optionTwo.equals("get")) {
            int hearts = (int) (targetPlayerData.getMaxHealth() / 2);
            sender.sendMessage(MessageUtils.getAndFormatMsg(true, "messages.getHearts", "&c%player% &7currently has &c%amount% &7hearts!",
                    new MessageUtils.Replaceable("%player%", targetPlayerName), new MessageUtils.Replaceable("%amount%", hearts + "")));
            return true;
        }

        int amount = Integer.parseInt(args[3]);

        if (amount < 0) {
            throwUsageError(sender, "/lifestealz hearts <add | set | remove> <player> [amount]");
            return false;
        }

        int finalAmount = amount;

        switch (optionTwo) {
            case "add": {
                if (config.getBoolean("enforceMaxHeartsOnAdminCommands") && targetPlayerData.getMaxHealth() + (amount * 2) > config.getInt("maxHearts") * 2) {
                    Component maxHeartsMsg = MessageUtils.getAndFormatMsg(true, "messages.maxHeartLimitReached", "&cYou already reached the limit of %limit% hearts!",
                            new MessageUtils.Replaceable("%limit%", config.getInt("maxHearts") + ""));
                    sender.sendMessage(maxHeartsMsg);
                    return false;
                }

                targetPlayerData.setMaxHealth(targetPlayerData.getMaxHealth() + (amount * 2));
                storage.save(targetPlayerData);
                LifeStealZ.setMaxHealth(targetPlayer, targetPlayerData.getMaxHealth());
                finalAmount = (int) (targetPlayerData.getMaxHealth() / 2);
                break;
            }
            case "set": {
                if (amount == 0) {
                    sender.sendMessage(Component.text("§cYou cannot set the lives below or to zero"));
                    return false;
                }

                if (config.getBoolean("enforceMaxHeartsOnAdminCommands") && amount > config.getInt("maxHearts")) {
                    Component maxHeartsMsg = MessageUtils.getAndFormatMsg(true, "messages.maxHeartLimitReached", "&cYou already reached the limit of %limit% hearts!",
                            new MessageUtils.Replaceable("%limit%", config.getInt("maxHearts") + ""));
                    sender.sendMessage(maxHeartsMsg);
                    return false;
                }

                targetPlayerData.setMaxHealth(amount * 2);
                storage.save(targetPlayerData);
                LifeStealZ.setMaxHealth(targetPlayer, targetPlayerData.getMaxHealth());
                break;
            }
            case "remove": {
                if ((targetPlayerData.getMaxHealth() / 2) - (double) amount <= 0) {
                    sender.sendMessage(Component.text("§cYou cannot set the lives below or to zero"));
                    return false;
                }

                targetPlayerData.setMaxHealth(targetPlayerData.getMaxHealth() - (amount * 2));
                storage.save(targetPlayerData);
                LifeStealZ.setMaxHealth(targetPlayer, targetPlayerData.getMaxHealth());
                finalAmount = (int) (targetPlayerData.getMaxHealth() / 2);
                break;
            }
        }

        Component setHeartsConfirmMessage = MessageUtils.getAndFormatMsg(true, "messages.setHeartsConfirm", "&7You successfully %option% &c%player%' hearts to &7%amount% hearts!",
                new MessageUtils.Replaceable("%option%", optionTwo), new MessageUtils.Replaceable("%player%", targetPlayerName), new MessageUtils.Replaceable("%amount%", finalAmount + ""));
        sender.sendMessage(setHeartsConfirmMessage);
        return true;
    }

    private boolean handleGiveItem(CommandSender sender, String[] args) {
        if (!sender.hasPermission("lifestealz.admin.giveitem")) {
            throwPermissionError(sender);
            return false;
        }

        if (args.length < 3) {
            throwUsageError(sender, "/lifestealz giveItem <player> <item> [amount]");
            return false;
        }

        String targetPlayerName = args[1];

        if (targetPlayerName == null) {
            throwUsageError(sender, "/lifestealz giveItem <player> <item> [amount]");
            return false;
        }

        Player targetPlayer = plugin.getServer().getPlayer(targetPlayerName);

        if (targetPlayer == null) {
            throwUsageError(sender, "/lifestealz giveItem <player> <item> [amount]");
            return false;
        }

        String item = args[2];

        if (item == null) {
            throwUsageError(sender, "/lifestealz giveItem <player> <item> [amount]");
            return false;
        }

        Set<String> possibleItems = plugin.getRecipeManager().getRecipeIds();

        if (!possibleItems.contains(item)) {
            throwUsageError(sender, "/lifestealz giveItem <player> <item> [amount]");
            return false;
        }

        int amount = args.length > 3 ? Integer.parseInt(args[3]) : 1;

        if (amount < 1) {
            throwUsageError(sender, "/lifestealz giveItem <player> <item> [amount]");
            return false;
        }

        boolean silent = args.length > 4 && args[4].equals("silent");

        targetPlayer.getInventory().addItem(CustomItemManager.createCustomItem(item, amount));
        if (!silent)
            targetPlayer.sendMessage(MessageUtils.getAndFormatMsg(true, "messages.giveItem", "&7You received &c%amount% &7%item%!",
                    new MessageUtils.Replaceable("%amount%", amount + ""), new MessageUtils.Replaceable("%item%", CustomItemManager.getCustomItemData(item).getName())));
        return true;
    }

    private boolean handleData(CommandSender sender, String[] args) {
        if (!sender.hasPermission("lifestealz.managedata")) {
            throwPermissionError(sender);
            return false;
        }

        if (args.length < 3) {
            throwUsageError(sender, "/lifestealz data <import | export> <file>");
            return false;
        }

        String optionTwo = args[1];
        String fileName = args[2];

        if (optionTwo.equals("export")) {
            storage.export(fileName);
            sender.sendMessage(MessageUtils.getAndFormatMsg(true, "messages.exportData", "&7Successfully exported player data to &c%file%.csv",
                    new MessageUtils.Replaceable("%file%", fileName)));
        } else if (optionTwo.equals("import")) {
            storage.importData(fileName);
            sender.sendMessage(MessageUtils.getAndFormatMsg(true, "messages.importData", "&7Successfully imported &c%file%.csv&7!\n&cPlease restart the server, to ensure flawless migration!",
                    new MessageUtils.Replaceable("%file%", fileName)));
        } else {
            throwUsageError(sender, "/lifestealz giveItem <player> <item> [amount]");
        }
        return true;
    }

    private boolean handleDev(CommandSender sender, String[] args) {
        if (!sender.isOp()) {
            throwPermissionError(sender);
            return false;
        }

        if (args.length < 2) {
            throwUsageError(sender, "/lifestealz dev <giveForbiddenitem>");
            return false;
        }

        String optionTwo = args[1];

        if (optionTwo.equals("giveForbiddenitem")) {
            if (!(sender instanceof Player)) return false;

            Player player = (Player) sender;

            player.getInventory().addItem(new CustomItem(Material.BARRIER).makeForbidden().getItemStack());

            return true;
        }

        return false;
    }

    private void throwUsageError(CommandSender sender, String usage) {
        Component msg = MessageUtils.getAndFormatMsg(false, "messages.usageError", "&cUsage: %usage%",
                new MessageUtils.Replaceable("%usage%", usage));
        sender.sendMessage(msg);
    }

    private void throwPermissionError(CommandSender sender) {
        Component msg = MessageUtils.getAndFormatMsg(false, "messages.noPermissionError", "&cYou don't have permission to use this!");
        sender.sendMessage(msg);
    }

    @Override
    public @Nullable List<String> onTabComplete(@NotNull CommandSender sender, @NotNull Command command, @NotNull String alias, String[] args) {
        if (args.length == 1) {
            List<String>
                    availableOptions = new ArrayList<>();
            if (sender.hasPermission("lifestealz.admin.reload")) availableOptions.add("reload");
            if (sender.hasPermission("lifestealz.admin.setlife")) availableOptions.add("hearts");
            if (sender.hasPermission("lifestealz.admin.giveitem")) availableOptions.add("giveItem");
            if (sender.hasPermission("lifestealz.viewrecipes")) availableOptions.add("recipe");
            if (sender.hasPermission("lifestealz.help")) availableOptions.add("help");
            if (sender.hasPermission("lifestealz.managedata")) availableOptions.add("data");
            return availableOptions;
        } else if (args.length == 2) {
            if (args[0].equals("hearts")) {
                return List.of("add", "set", "remove", "get");
            } else if (args[0].equals("giveItem")) {
                return null; // Suggest online player names
            } else if (args[0].equals("recipe")) {
                return new ArrayList<>(plugin.getRecipeManager().getRecipeIds());
            } else if (args[0].equals("data") && sender.hasPermission("lifestealz.managedata")) {
                return List.of("export", "import");
            } else if (args[0].equals("dev")) {
                return List.of("giveForbiddenitem");
            }
        } else if (args.length == 3) {
            if (args[0].equals("hearts")) {
                return null; // Suggest online player names
            } else if (args[0].equals("giveItem")) {
                return new ArrayList<>(plugin.getRecipeManager().getRecipeIds());
            } else if (args[0].equals("data") && args[1].equals("import") && sender.hasPermission("lifestealz.managedata")) {
                return getCSVFiles();
            }
        } else if (args.length == 4) {
            if (args[0].equals("hearts") || args[0].equals("giveItem")) {
                return List.of("1", "32", "64");
            }
        } else if (args.length == 5) {
            if (args[0].equals("giveItem")) {
                return List.of("silent");
            }
        }
        return null;
    }

    private List<String> getCSVFiles() {
        List<String> csvFiles = new ArrayList<>();
        File pluginFolder = plugin.getDataFolder();
        File[] files = pluginFolder.listFiles((dir, name) -> name.toLowerCase().endsWith(".csv"));
        if (files != null) {
            for (File file : files) {
                csvFiles.add(file.getName());
            }
        }
        return csvFiles;
    }
}

File: src/main/java/org/strassburger/lifestealz/commands/SettingsCommand.kt
package org.strassburger.lifestealz.commands

import net.kyori.adventure.text.Component
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.attribute.Attribute
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.ShapedRecipe
import org.bukkit.plugin.java.JavaPlugin
import org.strassburger.lifestealz.Lifestealz
import org.strassburger.lifestealz.util.ManageCustomItems
import org.strassburger.lifestealz.util.ManagePlayerdata
import org.strassburger.lifestealz.util.Replaceable

class SettingsCommand(private val plugin: JavaPlugin) : CommandExecutor {
    val config = Lifestealz.instance.config
    override fun onCommand(sender: CommandSender, command: Command, label: String, args: Array<out String>?): Boolean {

        val optionOne = args?.getOrNull(0)

        if (args == null || args.isEmpty() || optionOne == null) {
            sender.sendMessage(Lifestealz.getAndFormatMsg(true, "messages.versionMsg", "FALLBACK&7You are using version %version%", Replaceable("%version%", Lifestealz.instance.description.version)))
        }

        if (optionOne == "reload") {
            if (!sender.hasPermission("lifestealz.admin.reload")) {
                throwPermissionError(sender)
                return false
            }

            Lifestealz.instance.reloadConfig()

            Bukkit.resetRecipes()
            registerHeartRecipe()
            registerReviveRecipe()
            sender.sendMessage(Lifestealz.getAndFormatMsg(true, "messages.reloadMsg", "&7Successfully reloaded the plugin!"))
        }

        if (optionOne == "help") {
            if (!sender.hasPermission("lifestealz.help")) {
                throwPermissionError(sender)
                return false
            }

            var helpMessage: String = "&r \n&8----------------------------------------------------\n&c&lLifeStealZ &7help page\n&8----------------------------------------------------"
            helpMessage += "\n&c/lifestealz help &8- &7open this menu"
            if (sender.hasPermission("lifestealz.admin.reload")) helpMessage += "\n&c/lifestealz reload &8- &7reload the config"
            if (sender.hasPermission("lifestealz.admin.setlife")) helpMessage += "\n&c/lifestealz hearts &8- &7modify how many hearts a player has"
            if (sender.hasPermission("lifestealz.admin.giveitem")) helpMessage += "\n&c/lifestealz giveItem &8- &7give other players custom items, such as hearts"
            if (sender.hasPermission("lifestealz.admin.viewrecipes")) helpMessage += "\n&c/lifestealz recipe &8- &7view all recipes"
            if (sender.hasPermission("lifestealz.admin.revive")) helpMessage += "\n&c/revive &8- &7revive a player without a revive item"
            if (sender.hasPermission("lifestealz.admin.eliminate")) helpMessage += "\n&c/eliminate &8- &7eliminate a player"
            if (sender.hasPermission("lifestealz.withdraw")) helpMessage += "\n&c/withdrawheart &8- &7withdraw a heart"
            helpMessage += "\n&8----------------------------------------------------\n&r "

            val helpMessageFormatted = Lifestealz.formatMsg(helpMessage)

            sender.sendMessage(helpMessageFormatted)
        }

        if (optionOne == "recipe") {
            if (!sender.hasPermission("lifestealz.viewrecipes")) {
                throwPermissionError(sender)
                return false
            }

            if (sender !is Player) return false

            val availableRecipes = listOf("heart", "revivecrystal")

            val recipe = args.getOrNull(1)

            if (recipe == null || !availableRecipes.contains(recipe)) {
                throwRecipeUsageError(sender)
                return false
            }

            renderRecipe(player = sender, recipe = recipe)
        }

        if (optionOne == "hearts") {
            if (!sender.hasPermission("lifestealz.admin.setlife")) {
                throwPermissionError(sender)
                return false
            }

            val optionTwo = args.getOrNull(1)
            val possibleOptionTwo = listOf("add", "set", "remove", "get")

            if (optionTwo == null || !possibleOptionTwo.contains(optionTwo)) {
                throwUsageError(sender)
                return false
            }

            val targetPlayerName = args.getOrNull(2)

            if (targetPlayerName == null) {
                throwUsageError(sender)
                return false
            }

            val targetPlayer = Bukkit.getPlayer(targetPlayerName)

            if (targetPlayer == null) {
                throwUsageError(sender)
                return false
            }

            val targetplayerData = ManagePlayerdata().getPlayerData(uuid=targetPlayer.uniqueId.toString(), name=targetPlayer.name)

            if (optionTwo == "get") {
                val heartsString = (targetplayerData.maxhp / 2).toInt().toString()
                sender.sendMessage(Lifestealz.getAndFormatMsg(true, "messages.getHearts", "&c%player% &7currently has &c%amount% &7hearts!", Replaceable("%player%", targetPlayer.name), Replaceable("%amount%", heartsString)))
                return false
            }

            val amount = args.getOrNull(3)?.toIntOrNull()

            if (amount == null || amount < 0) {
                throwUsageError(sender)
                return false
            }

            if (optionTwo == "remove") {
                if ((targetplayerData.maxhp / 2) - amount.toDouble() <= 0) {
                    sender.sendMessage(Component.text("§cYou cannot set the lives below or to zero"))
                    return false
                }
            } else if (optionTwo == "set") {
                if (amount <= 0) {
                    sender.sendMessage(Component.text("§cYou cannot set the lives below or to zero"))
                    return false
                }
            }

            ManagePlayerdata().manageHearts(targetPlayer, optionTwo, amount.toDouble() * 2)

            when (optionTwo) {
                "add" -> {
                    if (config.getBoolean("enforceMaxHeartsOnAdminCommands") && targetPlayer.maxHealth + amount.toDouble() * 2 > config.getInt("maxHearts") * 2) {
                        val maxHeartsMsg = Lifestealz.getAndFormatMsg(true, "messages.maxHeartLimitReached", "&cYou already reached the limit of %limit% hearts!", Replaceable("%limit%", config.getInt("maxHearts").toString()))
                        sender.sendMessage(maxHeartsMsg)
                        return false
                    }

                    targetPlayer.maxHealth += amount.toDouble() * 2
                    targetPlayer.health += amount.toDouble() * 2
                }
                "set" -> {
                    if (config.getBoolean("enforceMaxHeartsOnAdminCommands") && amount > config.getInt("maxHearts")) {
                        val maxHeartsMsg = Lifestealz.getAndFormatMsg(true, "messages.maxHeartLimitReached", "&cYou already reached the limit of %limit% hearts!", Replaceable("%limit%", config.getInt("maxHearts").toString()))
                        sender.sendMessage(maxHeartsMsg)
                        return false
                    }
                    setMaxHealth(targetPlayer, amount.toDouble() * 2)
                }
                else -> {
                    targetPlayer.maxHealth -= amount.toDouble() * 2
                }
            }

            val setHeartsConfirmMessage = Lifestealz.getAndFormatMsg(true, "messages.setHeartsConfirm", "&7You successfully %option% &c%player%' hearts to &7%amount% hearts!", Replaceable("%option%", optionTwo), Replaceable("%player%", targetPlayer.name), Replaceable("%amount%", amount.toString()))
            sender.sendMessage(setHeartsConfirmMessage)

        }

        if (optionOne == "giveItem") {
            if (!sender.hasPermission("lifestealz.admin.giveitem")) {
                throwPermissionError(sender)
                return false
            }

            val targetPlayerName = args.getOrNull(1)

            if (targetPlayerName == null) {
                throwGiveItemUsageError(sender)
                return false
            }

            val targetPlayer = Bukkit.getPlayer(targetPlayerName)

            if (targetPlayer == null) {
                throwGiveItemUsageError(sender)
                return false
            }

            val item = args.getOrNull(2)
            val validItems = listOf("heart", "revivecrystal")

            if (item == null || !validItems.contains(item)) {
                throwGiveItemUsageError(sender)
                return false
            }

            var amount = args.getOrNull(3)?.toIntOrNull()

            if (amount == null) amount = 1

            when(item) {
                "heart" -> {
                    val heartItem = ManageCustomItems().createHeartItem()
                    heartItem.amount = amount
                    targetPlayer.inventory.addItem(heartItem)
                }

                "revivecrystal" -> {
                    val revivetItem = ManageCustomItems().createReviveItem()
                    revivetItem.amount = amount
                    targetPlayer.inventory.addItem(revivetItem)
                }

                else -> {}
            }

        }

        return true
    }

    private fun throwUsageError(sender: CommandSender) {
        val msg = Lifestealz.getAndFormatMsg(false, "messages.usageError", "&cUsage: %usage%", Replaceable("%usage%", "/lifestealz hearts <add | set | remove> <player> [amount]"))
        sender.sendMessage(msg)
    }

    private fun throwGiveItemUsageError(sender: CommandSender) {
        val msg = Lifestealz.getAndFormatMsg(false, "messages.usageError", "&cUsage: %usage%", Replaceable("%usage%", "/lifestealz giveItem <player> <item> [amount]"))
        sender.sendMessage(msg)
    }

    private fun throwRecipeUsageError(sender: CommandSender) {
        val msg = Lifestealz.getAndFormatMsg(false, "messages.usageError", "&cUsage: %usage%", Replaceable("%usage%", "/lifestealz recipe <heart | revivecrystal>"))
        sender.sendMessage(msg)
    }

    private fun throwPermissionError(sender: CommandSender) {
        val msg = Lifestealz.getAndFormatMsg(false, "messages.noPermissionError", "&cYou don't have permission to use this!")
        sender.sendMessage(msg)
    }

    fun setMaxHealth(player: Player, maxHealth: Double) {
        val attribute = player.getAttribute(Attribute.GENERIC_MAX_HEALTH)
        if (attribute != null) {
            attribute.baseValue = maxHealth
            player.health = maxHealth
        }
    }

    fun registerHeartRecipe() {
        if (!Lifestealz.instance.config.getBoolean("allowHeartCrafting")) return

        //Heart recipe
        val heartRecipeKey = NamespacedKey(plugin, "heartrecipe")
        val heartItem = ManageCustomItems().createHeartItem()
        val heartRecipe = ShapedRecipe(heartRecipeKey, heartItem)

        heartRecipe.shape("ABC", "DEF", "GHI")
        val rowOne = Lifestealz.instance.config.getStringList("items.heart.recipe.rowOne")
        val rowTwo = Lifestealz.instance.config.getStringList("items.heart.recipe.rowTwo")
        val rowThree = Lifestealz.instance.config.getStringList("items.heart.recipe.rowThree")
        heartRecipe.setIngredient('A', Material.valueOf(rowOne[0]))
        heartRecipe.setIngredient('B', Material.valueOf(rowOne[1]))
        heartRecipe.setIngredient('C', Material.valueOf(rowOne[2]))
        heartRecipe.setIngredient('D', Material.valueOf(rowTwo[0]))
        heartRecipe.setIngredient('E', Material.valueOf(rowTwo[1]))
        heartRecipe.setIngredient('F', Material.valueOf(rowTwo[2]))
        heartRecipe.setIngredient('G', Material.valueOf(rowThree[0]))
        heartRecipe.setIngredient('H', Material.valueOf(rowThree[1]))
        heartRecipe.setIngredient('I', Material.valueOf(rowThree[2]))

        Bukkit.addRecipe(heartRecipe)
    }

    fun registerReviveRecipe() {
        if (!Lifestealz.instance.config.getBoolean("allowReviveCrafting")) return

        //Revive crystal recipe
        val reviveRecipeKey = NamespacedKey(plugin, "reviverecipe")
        val reviveItem = ManageCustomItems().createReviveItem()
        val reviveRecipe = ShapedRecipe(reviveRecipeKey, reviveItem)

        reviveRecipe.shape("ABC", "DEF", "GHI")
        val reviverowOne = Lifestealz.instance.config.getStringList("items.revive.recipe.rowOne")
        val reviverowTwo = Lifestealz.instance.config.getStringList("items.revive.recipe.rowTwo")
        val reviverowThree = Lifestealz.instance.config.getStringList("items.revive.recipe.rowThree")
        reviveRecipe.setIngredient('A', Material.valueOf(reviverowOne[0]))
        reviveRecipe.setIngredient('B', Material.valueOf(reviverowOne[1]))
        reviveRecipe.setIngredient('C', Material.valueOf(reviverowOne[2]))
        reviveRecipe.setIngredient('D', Material.valueOf(reviverowTwo[0]))
        reviveRecipe.setIngredient('E', Material.valueOf(reviverowTwo[1]))
        reviveRecipe.setIngredient('F', Material.valueOf(reviverowTwo[2]))
        reviveRecipe.setIngredient('G', Material.valueOf(reviverowThree[0]))
        reviveRecipe.setIngredient('H', Material.valueOf(reviverowThree[1]))
        reviveRecipe.setIngredient('I', Material.valueOf(reviverowThree[2]))

        Bukkit.addRecipe(reviveRecipe)
    }

    private fun renderRecipe(player: Player, recipe: String) {
        val inventory: Inventory =  Bukkit.createInventory(null, 5 * 9, "§8Crafting recipe")

        inventory.setItem(40, ManageCustomItems().createCloseItem())

        val glass = ItemStack(Material.GRAY_STAINED_GLASS_PANE, 1)
        val glassMeta = glass.itemMeta
        glassMeta.displayName(Component.text("§r "))
        glass.itemMeta = glassMeta
        val glassSlots = listOf(0,1,2,3,4,5,6,7,8,9,13,14,15,16,17,18,22,23,25,26,27,31,32,33,34,35,36,37,38,39,41,42,43,44)
        for (slot in glassSlots) {
            inventory.setItem(slot, glass)
        }

        when(recipe) {
            "heart" -> {
                val rowOne = Lifestealz.instance.config.getStringList("items.heart.recipe.rowOne")
                val rowTwo = Lifestealz.instance.config.getStringList("items.heart.recipe.rowTwo")
                val rowThree = Lifestealz.instance.config.getStringList("items.heart.recipe.rowThree")
                inventory.setItem(10, ItemStack(Material.valueOf(rowOne[0]), 1))
                inventory.setItem(11, ItemStack(Material.valueOf(rowOne[1]), 1))
                inventory.setItem(12, ItemStack(Material.valueOf(rowOne[2]), 1))
                inventory.setItem(19, ItemStack(Material.valueOf(rowTwo[0]), 1))
                inventory.setItem(20, ItemStack(Material.valueOf(rowTwo[1]), 1))
                inventory.setItem(21, ItemStack(Material.valueOf(rowTwo[2]), 1))
                inventory.setItem(28, ItemStack(Material.valueOf(rowThree[0]), 1))
                inventory.setItem(29, ItemStack(Material.valueOf(rowThree[1]), 1))
                inventory.setItem(30, ItemStack(Material.valueOf(rowThree[2]), 1))

                inventory.setItem(24, ManageCustomItems().createHeartItem())
            }

            "revivecrystal" -> {
                val rowOne = Lifestealz.instance.config.getStringList("items.revive.recipe.rowOne")
                val rowTwo = Lifestealz.instance.config.getStringList("items.revive.recipe.rowTwo")
                val rowThree = Lifestealz.instance.config.getStringList("items.revive.recipe.rowThree")
                inventory.setItem(10, ItemStack(Material.valueOf(rowOne[0]), 1))
                inventory.setItem(11, ItemStack(Material.valueOf(rowOne[1]), 1))
                inventory.setItem(12, ItemStack(Material.valueOf(rowOne[2]), 1))
                inventory.setItem(19, ItemStack(Material.valueOf(rowTwo[0]), 1))
                inventory.setItem(20, ItemStack(Material.valueOf(rowTwo[1]), 1))
                inventory.setItem(21, ItemStack(Material.valueOf(rowTwo[2]), 1))
                inventory.setItem(28, ItemStack(Material.valueOf(rowThree[0]), 1))
                inventory.setItem(29, ItemStack(Material.valueOf(rowThree[1]), 1))
                inventory.setItem(30, ItemStack(Material.valueOf(rowThree[2]), 1))

                inventory.setItem(24, ManageCustomItems().createReviveItem())
            }
        }

        player.openInventory(inventory)
        Lifestealz.recipeGuiMap[player.uniqueId] = inventory
    }
}